source/extensions/filters/http/json_to_metadata/filter.cc

```diff
@@ -0,0 +1,391 @@
+  Rules rules;
+  for (const auto& rule : proto_config.request_rules()) {
+    rules.emplace_back(rule);
+  }
+  return rules;
+}
+
+absl::flat_hash_set<std::string> FilterConfig::generateRequestAllowContentTypes(
+    const envoy::extensions::filters::http::json_to_metadata::v3::JsonToMetadata& proto_config) {
+  if (proto_config.request_allow_content_types().empty()) {
+    return {Http::Headers::get().ContentTypeValues.Json};
+  }
+
+  absl::flat_hash_set<std::string> allow_content_types;
+  for (const auto& request_allowed_content_type : proto_config.request_allow_content_types()) {
+    allow_content_types.insert(request_allowed_content_type);
+  }
+  return allow_content_types;
+}
+
+bool FilterConfig::requestContentTypeAllowed(absl::string_view content_type) const {
+  if (content_type.empty()) {
+    return request_allow_empty_content_type_;
+  }
+
+  return request_allow_content_types_.contains(content_type);
+}
+
+void Filter::applyKeyValue(const std::string& value, const KeyValuePair& keyval,
+                           StructMap& struct_map) {
+  ASSERT(!value.empty());
+
+  ProtobufWkt::Value val;
+  val.set_string_value(value);
+  applyKeyValue(std::move(val), keyval, struct_map);
+}
+
+void Filter::applyKeyValue(double value, const KeyValuePair& keyval, StructMap& struct_map) {
+  ProtobufWkt::Value val;
+  val.set_number_value(value);
+  applyKeyValue(std::move(val), keyval, struct_map);
+}
+
+void Filter::applyKeyValue(ProtobufWkt::Value value, const KeyValuePair& keyval,
+                           StructMap& struct_map) {
+  const auto& nspace = decideNamespace(keyval.metadata_namespace());
+  addMetadata(nspace, keyval.key(), std::move(value), keyval.preserve_existing_metadata_value(),
+              struct_map);
+}
+
+const std::string& Filter::decideNamespace(const std::string& nspace) const {
+  static const std::string& jsonToMetadata = "envoy.filters.http.json_to_metadata";
+  return nspace.empty() ? jsonToMetadata : nspace;
+}
+
+bool Filter::addMetadata(const std::string& meta_namespace, const std::string& key,
+                         ProtobufWkt::Value val, const bool preserve_existing_metadata_value,
+                         StructMap& struct_map) {
+
+  if (preserve_existing_metadata_value) {
+    // TODO(kuochunghsu): support encoding
+    auto& filter_metadata = decoder_callbacks_->streamInfo().dynamicMetadata().filter_metadata();
+    const auto entry_it = filter_metadata.find(meta_namespace);
+
+    if (entry_it != filter_metadata.end()) {
+      const auto& metadata = entry_it->second;
+      if (metadata.fields().contains(key)) {
+        ENVOY_LOG(trace, "Found key {} in namespace {}. Preserve the existing metadata value.", key,
+                  meta_namespace);
+        return false;
+      }
+    }
+  }
+
+  ENVOY_LOG(trace, "add metadata ns:{} key:{}", meta_namespace, key);
+
+  // Have we seen this namespace before?
+  auto namespace_iter = struct_map.find(meta_namespace);
```

Comment:
No need to do the double find, we can just do 
`auto& keyval = struct_map[meta_namespace]` 

which should insert the default ctor if key not exist or provide you the value if it does.